File: lib/caesar.ex
defmodule Caesar do
  def start(_type, _args) do
    IO.puts decrypt("Hgx Angwkxw Hkvatkw Khtw, Vhgvhkwx Ahmxe & Lahiibgz Ftee, #sxkhmph-yhkmrmakxx", 19)
    {:ok, self()}
  end

  def encrypt(string, amount), do: shift(string, amount)
  def decrypt(string, amount), do: shift(string, -amount)

  defp shift(string, amount) when is_binary(string),
    do: for <<char <- string>>, into: <<>>, do: <<shift(char, amount)>>

  defp shift(char, amount) when char in ?A..?Z,
    do: shift(char, amount, ?A)
  defp shift(char, amount) when char in ?a..?z,
    do: shift(char, amount, ?a)
  defp shift(char, _amount),
    do: char

  defp shift(char, amount, base),
    do: Integer.mod(char + amount - base, 26) + base
end

File: src/main/kotlin/Caesar.kt
package com.jayjun.caesar

class Caesar(val amount: Int) {
    fun encrypt(string: String): String = shift(string, amount)
    fun decrypt(string: String): String = shift(string, -amount)

    fun shift(string: String, amount: Int): String {
        var result: String = ""
        for (character in string) {
            when (character) {
                in 'A'..'Z' -> result += shift(character, amount, 'A')
                in 'a'..'z' -> result += shift(character, amount, 'a')
                else -> result += character
            }
        }
        return result
    }

    fun shift(character: Char, amount: Int, base: Char): Char {
        var shift = (character + amount - base) % 26
        if (shift < 0) {
            shift += 26
        }
        return base + shift
    }
}

fun main(args: Array<String>) {
    val cipher = Caesar(19)
    println(cipher.decrypt("Hgx Angwkxw Hkvatkw Khtw, Vhgvhkwx Ahmxe & Lahiibgz Ftee, #sxkhmph-yhkmrmakxx"))
}
